packages/vite/src/node/build.ts

```diff
@@ -1092,7 +1092,10 @@ export function injectEnvironmentToHooks(
 
   const clone = { ...plugin }
 
-  for (const hook of Object.keys(clone) as RollupPluginHooks[]) {
+  for (const untypedHook in clone) {
+    // TypeScript automatically infers the keys of objects as `string`, so we need to explicitly type it.
+    // Otherwise, dependent type checks will fail
+    const hook = untypedHook as Extract<keyof typeof clone, RollupPluginHooks>
```

Comment:
I guess this is the only place inherited properties matters. In this case, I think this is more a fix and something we should do.

Reply:
According to the docs, "The Object.keys() static method returns an array of a given object's own enumerable string-keyed property names."

For..in also iterates over an object's enumerate keys, so the functionality is the same

Reply:
No, the functionality is slightly different.
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...in#iterating_over_own_properties_only